class.c

```diff
@@ -1364,8 +1378,23 @@ class_descendants_recursive(VALUE klass, VALUE ary)
 VALUE
 rb_class_descendants(VALUE klass)
 {
-    VALUE ary = rb_ary_new();
-    rb_class_foreach_subclass(klass, class_descendants_recursive, ary);
+    struct subclass_traverse_data data = { NULL, 0, -1 };
+
+    // estimate the count of subclasses
+    rb_class_foreach_subclass(klass, class_descendants_recursive, (VALUE) &data);
+
+    // this allocation may cause GC which may reduce the subclasses
+    data.buffer = ALLOC_N(VALUE, data.count);
```

Comment:
Can't also `ALLOC_N` trigger GC?

Reply:
Yes, that's what the above comment states.